File: src/decoders/pronote-date.ts
import { UnreachableError } from "~/models";

const SHORT_DATE_RE = /^\d{2}\/\d{2}\/\d{4}$/;
const LONG_DATE_LONG_HOURS_RE = /^\d{2}\/\d{2}\/\d{4} \d{2}:\d{2}:\d{2}$/;
const LONG_DATE_SHORT_HOURS_RE = /^\d{2}\/\d{2}\/\d{2} \d{2}h\d{2}$/;
const YEAR_FIRST_TWO_CHARS = new Date().getFullYear().toString().slice(0, 2);

export const decodePronoteDate = (formatted: string): Date => {
  if (SHORT_DATE_RE.test(formatted)) {
    const [day, month, year] = formatted.split("/").map(Number);
    return new Date(year, month - 1, day);
  }

  else if (LONG_DATE_LONG_HOURS_RE.test(formatted)) {
    const [date, time] =  formatted.split(" ");
    const [day, month, year] = date.split("/").map(Number);
    const [hours, minutes, seconds] = time.split(":").map(Number);

    const output = new Date(year, month - 1, day);
    output.setHours(hours, minutes, seconds);
    return output;
  }

  else if (LONG_DATE_SHORT_HOURS_RE.test(formatted)) {
    const [date, time] =  formatted.split(" ");
    const [day, month, year] = date.split("/").map(Number);
    const [hours, minutes] = time.split("h").map(Number);

    const output = new Date(parseInt(`${YEAR_FIRST_TWO_CHARS}${year}`), month - 1, day);
    output.setHours(hours, minutes);
    return output;
  }

  throw new UnreachableError("decodePronoteDate");
};

File: library/src/commonMain/kotlin/ink/literate/pawnote/decoders/pronote-date.kt
package ink.literate.pawnote.decoders

import ink.literate.pawnote.models.errors.UnreachableError

import kotlinx.datetime.*

const val shortDateRe = """^\d{2}\/\d{2}\/\d{4}$"""
const val longDateLongHoursRe = """^\d{2}\/\d{2}\/\d{4} \d{2}:\d{2}:\d{2}$"""
const val longDateShortHoursRe = """^\d{2}\/\d{2}\/\d{2} \d{2}h\d{2}$"""
val yearFirstTwoChars = Instant.fromEpochMilliseconds(Clock.System.now().toEpochMilliseconds()).toLocalDateTime(TimeZone.currentSystemDefault()).year.toString().slice(0..<2)

fun decodePronoteDate (formatted: String): LocalDateTime {
    if (shortDateRe.toRegex().containsMatchIn(formatted)) {
        val (day, month, year) = formatted.split("/").map { it.toInt() }
        return LocalDateTime(year, month, day, 0, 0)
    }
    else if (longDateLongHoursRe.toRegex().containsMatchIn(formatted)) {
        val (date, time) = formatted.split(" ")
        val (day, month, year) = date.split("/").map { it.toInt() }
        val (hours, minutes, seconds) = time.split(":").map { it.toInt() }

        return LocalDateTime(year, month, day, hours, minutes, seconds)
    }
    else if (longDateShortHoursRe.toRegex().containsMatchIn(formatted)) {
        val (date, time) = formatted.split(" ")
        val (day, month, year) = date.split("/").map { it.toInt() }
        val (hours, minutes) = time.split("h").map { it.toInt() }

        return LocalDateTime("${yearFirstTwoChars}${year}".toInt(), month, day, hours, minutes)
    }

    throw UnreachableError("decodePronoteDate")
}